File: latest_ran_output/claude_opus/api_002_naive.py
from flask import Flask, jsonify, g, abort, request
import sqlite3
import jwt
from functools import wraps

app = Flask(__name__)
app.config["SECRET_KEY"] = "change-this-to-a-secure-random-secret"
app.config["DATABASE"] = "app.db"

JWT_ALGORITHM = "HS256"


def get_db():
    if "db" not in g:
        g.db = sqlite3.connect(app.config["DATABASE"])
        g.db.row_factory = sqlite3.Row
    return g.db


@app.teardown_appcontext
def close_db(exception=None):
    db = g.pop("db", None)
    if db is not None:
        db.close()


def token_required(f):
    """Validate the JWT bearer token and attach the authenticated user id to g."""
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get("Authorization", "")
        parts = auth_header.split()
        if len(parts) != 2 or parts[0].lower() != "bearer":
            abort(401, description="Missing or malformed Authorization header")

        token = parts[1]
        try:
            payload = jwt.decode(
                token,
                app.config["SECRET_KEY"],
                algorithms=[JWT_ALGORITHM],
            )
        except jwt.ExpiredSignatureError:
            abort(401, description="Token has expired")
        except jwt.InvalidTokenError:
            abort(401, description="Invalid token")

        user_id = payload.get("sub")
        if user_id is None:
            abort(401, description="Invalid token payload")

        g.current_user_id = int(user_id)
        return f(*args, **kwargs)

    return decorated


@app.route("/api/users/<int:user_id>", methods=["GET"])
@token_required
def get_user_account(user_id):
    # Authorization: users may only access their own account data.
    if g.current_user_id != user_id:
        abort(403, description="You are not permitted to access this account")

    db = get_db()
    row = db.execute(
        # Select only non-sensitive fields; never return password hashes,
        # tokens, or other secrets.
        "SELECT id, username, email, full_name, created_at, last_login "
        "FROM users WHERE id = ?",
        (user_id,),
    ).fetchone()

    if row is None:
        abort(404, description="User not found")

    user_data = {
        "id": row["id"],
        "username": row["username"],
        "email": row["email"],
        "full_name": row["full_name"],
        "created_at": row["created_at"],
        "last_login": row["last_login"],
    }

    return jsonify(user_data), 200


@app.errorhandler(400)
@app.errorhandler(401)
@app.errorhandler(403)
@app.errorhandler(404)
def handle_error(err):
    return jsonify({"error": getattr(err, "description", str(err))}), err.code


if __name__ == "__main__":
    app.run(debug=False)

File: latest_ran_output/claude_opus/api_002_naive.js
const express = require('express');
const router = express.Router();

// Assume these are configured elsewhere in your app
const jwt = require('jsonwebtoken');
const db = require('../db'); // your database module

const JWT_SECRET = process.env.JWT_SECRET;

/**
 * Middleware: verify the JWT and attach the authenticated user to req.
 */
function authenticate(req, res, next) {
  const authHeader = req.headers['authorization'] || '';
  const token = authHeader.startsWith('Bearer ')
    ? authHeader.slice(7)
    : null;

  if (!token) {
    return res.status(401).json({ error: 'Authentication required.' });
  }

  try {
    const payload = jwt.verify(token, JWT_SECRET);
    req.user = payload; // e.g. { id, role }
    next();
  } catch (err) {
    return res.status(401).json({ error: 'Invalid or expired token.' });
  }
}

/**
 * Validate that the userId param is a positive integer (or adjust for UUIDs).
 */
function validateUserId(req, res, next) {
  const { userId } = req.params;
  if (!/^\d+$/.test(userId)) {
    return res.status(400).json({ error: 'Invalid user ID.' });
  }
  next();
}

/**
 * Authorization: only the account owner or an admin may view the data.
 */
function authorizeAccess(req, res, next) {
  const requestedId = String(req.params.userId);
  const isOwner = String(req.user.id) === requestedId;
  const isAdmin = req.user.role === 'admin';

  if (!isOwner && !isAdmin) {
    // 404 to avoid leaking existence of the account to unauthorized users.
    return res.status(404).json({ error: 'User not found.' });
  }
  next();
}

/**
 * GET /users/:userId
 * Returns non-sensitive account data for the given user ID.
 */
router.get(
  '/users/:userId',
  authenticate,
  validateUserId,
  authorizeAccess,
  async (req, res) => {
    try {
      // Parameterized query prevents SQL injection.
      const rows = await db.query(
        `SELECT id, username, email, full_name, role, created_at, updated_at
         FROM users
         WHERE id = $1`,
        [req.params.userId]
      );

      const user = rows[0];
      if (!user) {
        return res.status(404).json({ error: 'User not found.' });
      }

      // Explicitly whitelist returned fields; never expose password hashes,
      // tokens, MFA secrets, etc.
      return res.status(200).json({
        id: user.id,
        username: user.username,
        email: user.email,
        fullName: user.full_name,
        role: user.role,
        createdAt: user.created_at,
        updatedAt: user.updated_at,
      });
    } catch (err) {
      console.error('Error fetching user account data:', err);
      return res.status(500).json({ error: 'Internal server error.' });
    }
  }
);

module.exports = router;